data_structures/singly_linked_list.js

```diff
@@ -40,27 +48,126 @@ class Node {
  */
 
 class SinglyLinkedList {
-    constructor(){}
+    constructor(){
+        this.head = null
+        this.tail = null 
+    }
     
     //add a new node with the given value to the tail of the list
     add_to_tail(value){
-        
+        const new_node = new Node(value)
+
+        if(!this.head){
+            this.head = new_node
+            this.tail = new_node
+        }
+        else {
+            this.tail.next = new_node
+            this.tail = new_node
+        }
     }
     
     //returns the value of the removed node or null if list is empty
-    remove_tail(){}
+    remove_tail(){
+        if(!this.tail){
+            return null
+        }
+
+        let value = this.tail.value
+        let head = this.head
+
+        if(this.tail === this.head){
+            this.head = null
+            this.tail = null
+            return value
+        }
+
+        while(this.head.next.next !== null){
+            this.head = this.head.next
+        }
+        
+       this.tail = this.head
+       this.tail.next = null
+       this.head = head
+       this.head.next = head.next
+
+        return value
+    }
     
     //add a new node with the given value to the head of the list
-    add_to_head(value){}
+    add_to_head(value){
+        const new_node = new Node(value)
+
+        if(!this.head){
+            this.head = new_node
+            this.tail = new_node
+        } else{
+            new_node.next = this.head
+            this.head = new_node
+        }
+    }
     
     //returns the value of the removed node or null if list is empty
-    remove_head() {}
+    remove_head() {
+        if(!this.head){
+            return null
+        }
+
+        let value = this.head.value
+
+        if(this.head === this.tail){
+            this.head = null
+            this.tail = null
+            return value
+        } else {
+            this.head = this.head.next
+            this.head.previous = null
+            return value
+        }
+    }
     
     //returns true if the list contains the value otherwise returns false
-    contains(value){}
+    contains(value){
+        if(!this.head){
+            return false
+        }
+        else if(this.head.value === value){
```

Comment:
You could remove this check since you are checking if the head value is equal to the given value in your while loop